src/nvim/edit.c

```diff
@@ -1375,13 +1375,19 @@ ins_redraw (
 
   // Trigger TextChangedI if b_changedtick differs.
   if (ready && has_event(EVENT_TEXTCHANGEDI)
-      && last_changedtick != curbuf->b_changedtick
+      && curbuf->b_last_changedtick != curbuf->b_changedtick
       && !pum_visible()) {
-    if (last_changedtick_buf == curbuf) {
       apply_autocmds(EVENT_TEXTCHANGEDI, NULL, NULL, false, curbuf);
-    }
-    last_changedtick_buf = curbuf;
-    last_changedtick = curbuf->b_changedtick;
+      curbuf->b_last_changedtick = curbuf->b_changedtick;
+  }
+
+  // Trigger TextChangedP if b_changedtick differs. When the popupmenu closes
+  // TextChangedI will need to trigger for backwards compatibility, thus use
+  // different b_last_changedtick* variables.
+  if (ready && curbuf->b_last_changedtick_pum != curbuf->b_changedtick
+          && pum_visible()) {
```

Comment:
why was the check for `has_textchangedP()` not included?

Please don't silently make changes to ported patches without explaining why. It makes it really difficult to trust that the patches are being ported correctly.

Reply:
`has_textchangedP()` implementation uses `autopat` feature.
It is not included in neovim.
And it just checks auto command is defined.

Reply:
We have generic `has_event`, it should be used instead.

Reply:
OK.   I will fix it.